Added framework upgrade page. (#16)
SkillsFundingAgency/das-technical-guidance: collections/_development/framework-upgrade-roadmap.md

---
layout: default
title:  "Framework Upgrades"
category: development
---

## Framework Upgrades

> There is always a clamour to utilise the latest versions of the frameworks we use to develop applications but it is important that a process is followed to ensure we don't increase the burden of supporting multiple different framework versions unnecessarily.  We should be in a position where by applications which are actively being developed are upgraded to the latest framework version as a matter of course.

Reviewer: Not sure about the working here. Maybe more like:
```suggestion
There is always the temptation to...
```

### New solutions

> New solutions should always use the latest version of the framework which is in Long Term Support.

Reviewer: I think a broader explanation as to why we only want to use LTS will help to head off any questions

### When to upgrade

> * As soon as an application is undergoing significant development

Reviewer: ```suggestion
* When an application is undergoing significant development
```

* When there is a team in place to support testing of the upgrade
* When the new version of the framework is no longer in preview

### When not to upgrade

> * When the new version is not covered by Long Term Support.  Details of the versions covered by LTS can be found on the [Microsoft website](https://dotnet.microsoft.com/platform/support/policy/dotnet-core)

Reviewer: ```suggestion
* When the new framework version is not covered by Long Term Support. Details of the versions covered by LTS can be found on the [Microsoft website](https://dotnet.microsoft.com/platform/support/policy/dotnet-core)
```

> * When the application is dormant or changes are very minor and infrequent

Reviewer: ```suggestion
* When the application is in the maintenance phase of its life-cycle, or when changes to it are very minor and infrequent
```

* When the application is due to be decommissioned

### Upgrade process

> * Speak to Dev Ops - this first step in the process should always be to engage the Dev Ops team.  If you're the first person to implement the new version then additional work may be required for the Dev Ops team which will need to be planned.

Reviewer: ```suggestion
* Engage with DevOps to discuss the upgrade. The first time that a new version is used there may be additional work  required by the DevOps team to upgrade build servers for example that will need to be planned into their workload.
```

> * Agree the work with your PO and DM.  If they aren't supportive then spec to either the Dev Lead, Tech Lead or Quality Guild.

Reviewer 2: spelling :-)

Author: I love that your first git contribution in 6 months is to pick me up on my dodgy spelling. 😂

Reviewer 2: I'm sorry I couldn't resist 😆 
Hope everything is going well.

> * Agree the work with your PO and DM.  If they aren't supportive then speak to either the Dev Lead, Tech Lead or Quality Guild.

Reviewer: ```suggestion
* Agree the extra work to upgrade with your Delivery Team. If they are not supportive ask them to speak to the Developer Clan Lead to discuss their objections. It may then be raised with the Quality Guild to gain wider opinions if necessary.
```

* Upgrade the solution
* Liase with Dev Ops to undertake any work required on the build/release pipelines

> * Test thoroughly locally

Reviewer: ```suggestion
* Thoroughly test the solution in your local environment, ensuring all unit tests and any integration tests pass
```

> * Execute thorough regression testing

Reviewer: ```suggestion
* Ensure that the solution is thoroughly regression tested
```
